### PR TITLE
Script for generating S3 versioned file stats

### LINKED ISSUE #470: tools/s3-gc-stats helper script
dandiarchive bucket has versioning turned on, which is a great feature to guarantee access to the specific version of a key in case it mutates.  Versioning though brings in potential overhead, since deleted keys are not really deleted but rather DeleteMarker key is placed as the last "version" of that key.  Before we jump into GC of the bucket to get rid of no longer active prior versions, we need to be able to assess the current state of the given folder/prefix within s3 bucket.  

So I think we need a script like

`s3-gc-stats --stat ["all","visible","invisible", "old"] [--list] [--fail-if-any] bucket [prefix]` 

which would traverse all versions of all the keys and provide stats on 
how many (number)/much (total size) (and a list of them if ran with `--list`) for following cases ("all" by default)

- `visible`: files/keys are "visible" (not DeleteMarker on top) 
- `invisible`: keys are "invisible" (as files) -- correspond to files/prefixes that were either removed (DeleteMarker is their last version) or superseded by a newer version; note: same file (prefix) could have multiple invisible keys
- `old`: keys are "old" - correspond to prior versions of still valid files (so there is no DeleteMarker on top)

<details>
<summary>`s3://datalad-test0-versioned` might be a nice tiny sample bucket to try on since I believe it has a good number of use cases covered:</summary> 

```shell
$> datalad ls -La -r s3://datalad-test0-versioned
Connecting to bucket: datalad-test0-versioned
[INFO   ] S3 session: Connecting to the bucket datalad-test0-versioned with authentication 
Bucket info:
  Versioning: {'Versioning': 'Enabled', 'MfaDelete': 'Disabled'}
     Website: datalad-test0-versioned.s3-website-us-east-1.amazonaws.com
         ACL: <Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>
1version-nonversioned1.txt                 2015-11-07T05:23:36.000Z 8 ver:null                              acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/1version-nonversioned1.txt?versionId=null [OK]
1version-removed-recreated.txt             2015-11-07T05:23:38.000Z 8 ver:vurmCZo5d3.iAzTV.YAfTcMr.LDVjQIK  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/1version-removed-recreated.txt?versionId=vurmCZo5d3.iAzTV.YAfTcMr.LDVjQIK [OK]
1version-removed-recreated.txt             2015-11-07T05:23:38.000Z DeleteMarker
1version-removed-recreated.txt             2015-11-07T05:23:38.000Z 8 ver:B7cP2PKGRNeRZ7EktWN82UnUaGjb9dh_  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/1version-removed-recreated.txt?versionId=B7cP2PKGRNeRZ7EktWN82UnUaGjb9dh_ [OK]
1version-removed.txt                       2015-11-07T05:23:37.000Z DeleteMarker
1version-removed.txt                       2015-11-07T05:23:37.000Z 8 ver:eZ5Hgwo8azfBv3QT7aW9dmm2sbLUY.QP  acl:S3ResponseError: 404 Not Found
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Key>1version-removed.txt</Key><RequestId>FCRBDCVKNVMY1MBJ</RequestId><HostId>AB1OdmQw4GWXfOK0h9ZcMEcNeh20Li/vpQz80fAn1/thA100x7OzjdveWSJq7mriR3TYovlisvs=</HostId></Error>  http://datalad-test0-versioned.s3.amazonaws.com/1version-removed.txt?versionId=eZ5Hgwo8azfBv3QT7aW9dmm2sbLUY.QP [OK]
2versions-nonversioned1.txt                2015-11-07T05:23:37.000Z 8 ver:V4Dqhu0QTEtxmvoNkCHGrjVZVomR1Ryo  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/2versions-nonversioned1.txt?versionId=V4Dqhu0QTEtxmvoNkCHGrjVZVomR1Ryo [OK]
2versions-nonversioned1.txt                2015-11-07T05:23:36.000Z 8 ver:null                              acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/2versions-nonversioned1.txt?versionId=null [OK]
2versions-nonversioned1.txt_sameprefix     2015-11-07T05:23:37.000Z 8 ver:QUOLbSrBHBmU3UNeEvdfcX2puqRffH9l  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/2versions-nonversioned1.txt_sameprefix?versionId=QUOLbSrBHBmU3UNeEvdfcX2puqRffH9l [OK]
2versions-removed-recreated.txt            2015-11-07T05:23:37.000Z 8 ver:bBVSCB4MdBOeEXDQ2KwrjtrevpwFabaY  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/2versions-removed-recreated.txt?versionId=bBVSCB4MdBOeEXDQ2KwrjtrevpwFabaY [OK]
2versions-removed-recreated.txt            2015-11-07T05:23:37.000Z DeleteMarker
2versions-removed-recreated.txt            2015-11-07T05:23:37.000Z 8 ver:zwW0b567gYO3puJeLMZsOmETqowJnv6l  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/2versions-removed-recreated.txt?versionId=zwW0b567gYO3puJeLMZsOmETqowJnv6l [OK]
2versions-removed-recreated.txt_sameprefix 2015-11-07T05:23:37.000Z 8 ver:cfTdf3N8exZLFg.KcW5szQKrFNLUyCu1  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/2versions-removed-recreated.txt_sameprefix?versionId=cfTdf3N8exZLFg.KcW5szQKrFNLUyCu1 [OK]
3versions-allversioned.txt                 2015-11-07T05:23:37.000Z 8 ver:Kvuind11HZh._dCPaDAb0OY9dRrQoTMn  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/3versions-allversioned.txt?versionId=Kvuind11HZh._dCPaDAb0OY9dRrQoTMn [OK]
3versions-allversioned.txt                 2015-11-07T05:23:37.000Z 8 ver:b.qCuh7Sg58VIYj8TVHzbRS97EvejzEl  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/3versions-allversioned.txt?versionId=b.qCuh7Sg58VIYj8TVHzbRS97EvejzEl [OK]
3versions-allversioned.txt                 2015-11-07T05:23:37.000Z 8 ver:pNsV5jJrnGATkmNrP8.i_xNH6CY4Mo5s  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/3versions-allversioned.txt?versionId=pNsV5jJrnGATkmNrP8.i_xNH6CY4Mo5s [OK]
3versions-allversioned.txt_sameprefix      2015-11-07T05:23:37.000Z 8 ver:Mvsc4FgJWc6gExwSw1d6wsLrnk6wdDVa  acl:<Policy: yoh@cs.unm.edu (owner) = FULL_CONTROL>  http://datalad-test0-versioned.s3.amazonaws.com/3versions-allversioned.txt_sameprefix?versionId=Mvsc4FgJWc6gExwSw1d6wsLrnk6wdDVa [OK]
```
</details>

immediate target use cases will be 
- `s3://dandiarchive/` as a whole so we could assess how many skeletons are in the closet
- `--stat "old" --list --fail-if-any s3://dandiarchive/blobs` (as soon as we get new dandi-api blob store in the public bucket) - to be alerted to what should in theory not happen since blobs should be immutable and there should be no multiple versions of the same blob (or blobs with colliding UUIDs) 

@satra - any other target metric we might like/need?

edit: could as well just be a `DANDI_DEVEL` only available command of `dandi` client and not a separate script.

### PR DESCRIPTION
**Pull request by jwodder**
Closes #470.

### CODE REVIEW

**Comment by yarikoptic** [comment]
Please `humanize.naturalsize` sizes, e.g.
```
Invisible files: 14241 / Size: 8104409130145 (8.1 TB)
```

**Comment by yarikoptic** [comment]
FTR, here is what I have got using 0.11.0-14-g3d246b8 version for s3://dandiarchive:

```shell
$> grep Size /tmp/dandiarchive-*
/tmp/dandiarchive-all.stat:All files: 197558 / Size: 26795152820245
/tmp/dandiarchive-invisible.stat:Invisible files: 14241 / Size: 8104409130145
/tmp/dandiarchive-old.stat:Old files: 538 / Size: 68846538
/tmp/dandiarchive-visible.stat:Visible files: 182779 / Size: 18690674843562
```
yet to analyze if kosher (invisible of 8TB sounds excessive but may be it is life)

**Comment by yarikoptic** [comment]
for invisible there is a good number of such as
```
s3://dandiarchive/dandiarchive/dandiarchive/data/?versionId=xRkqMiJ_F_OwegaT_m8dtI1MQW6B7SF3 0
s3://dandiarchive/dandiarchive/dandiarchive/data/?versionId=TN8czMbM5omaaeuwEtSaarGsH6lOz.io 0
```
which are a bit odd to me -- seems to be "directories" due to having trailing `/` in them, did not know that key could exist like that... datalad fails to "download" them with 403:

```shell
$> datalad download-url s3://dandiarchive/dandiarchive/dandiarchive/data/?versionId=QHPjluoj8pQm0y7Lau7ldEiQ7rEiQXy5
[INFO   ] Downloading 's3://dandiarchive/dandiarchive/dandiarchive/data/?versionId=QHPjluoj8pQm0y7Lau7ldEiQ7rEiQXy5' into '/tmp/' 
[INFO   ] S3 session: Connecting to the bucket dandiarchive anonymously 
download_url(error): /tmp/ (file) [S3 refused to provide the key for dandiarchive/dandiarchive/data/ from url s3://dandiarchive/dandiarchive/dandiarchive/data/?versionId=QHPjluoj8pQm0y7Lau7ldEiQ7rEiQXy5: S3ResponseError: 403 Forbidden
 [s3.py:get_downloader_session:352]]
```

anyways -- I see that there was a fix added 3d246b899db3bf96bdd2c87136a6bcc754d3e94a so will rerun now to get a cleaner picture

**Comment by satra** [comment]
we should also skip that directory: `s3://dandiarchive/dandiarchive` that's added as a private component related to AWS inventory function and is non public i believe.

**Comment by yarikoptic** [comment]
@jwodder could you please add `--exclude URL_REGEX` so we could exclude some URLs?

**Comment by jwodder** [comment]
@yarikoptic Is this for matching against just URLs of the form `s3://BUCKET/KEY`, or should versionId's be included in the URLs being matched against?

**Comment by yarikoptic** [comment]
current use case is just the key (prefix) itself, not vesionIds which would be "fun" to match pragmatically via regexes anyways.

**Comment by jwodder** [comment]
@yarikoptic `--exclude` option added.

File: tools/s3-gc-stats [hunk 1]
```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python3
+
+# Assumptions made by this code:
+# - list_object_versions sorts all versions (including delete markers) by key
+#   (in ascending order) and last modified date (in descending order, sometimes
+#   with consecutive equal timestamps) and returns them in chunks of 1000 (by
+#   default), with each chunk divided into proper versions and delete markers.
```
Comment (yarikoptic) [inline tools/s3-gc-stats:7]: note: these assumptions seems to be validated in the code below: https://github.com/dandi/dandi-cli/pull/473/files#diff-c840778a73d440599aee9be998e0b555e84831c627e589f54439e8d3f2819295R103 . Great!

File: tools/s3-gc-stats [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        # Use s3 anonymously/without credentials:
+        client = boto3.client("s3", config=Config(signature_version=UNSIGNED))
+        for page in client.get_paginator("list_object_versions").paginate(
+            Bucket=self.bucket, Prefix=self.prefix
+        ):
+            # TODO: Filter out keys that ends with slashes?
+            for dm in page.get("DeleteMarkers", []):
+                if dm["IsLatest"]:
+                    self.mark_deleted(dm["Key"])
+            for v in page["Versions"]:
```
Comment (yarikoptic) [inline tools/s3-gc-stats:91]: <details>
<summary>note: it seems to correctly work out the keys which were added before versioning was turned on
 -- look at 1version-nonversioned1.txt and 2versions-nonversioned1.txt . GREAT!</summary> 

```shell
(git)lena:~/proj/dandi/dandi-cli-master[gh-470]git
$> tools/s3-gc-stats --stat visible --list s3://datalad-test0-versioned
s3://datalad-test0-versioned/1version-nonversioned1.txt?versionId=null 8
s3://datalad-test0-versioned/1version-removed-recreated.txt?versionId=vurmCZo5d3.iAzTV.YAfTcMr.LDVjQIK 8
s3://datalad-test0-versioned/2versions-nonversioned1.txt?versionId=V4Dqhu0QTEtxmvoNkCHGrjVZVomR1Ryo 8
s3://datalad-test0-versioned/2versions-nonversioned1.txt_sameprefix?versionId=QUOLbSrBHBmU3UNeEvdfcX2puqRffH9l 8
s3://datalad-test0-versioned/2versions-removed-recreated.txt?versionId=bBVSCB4MdBOeEXDQ2KwrjtrevpwFabaY 8
s3://datalad-test0-versioned/2versions-removed-recreated.txt_sameprefix?versionId=cfTdf3N8exZLFg.KcW5szQKrFNLUyCu1 8
s3://datalad-test0-versioned/3versions-allversioned.txt?versionId=Kvuind11HZh._dCPaDAb0OY9dRrQoTMn 8
s3://datalad-test0-versioned/3versions-allversioned.txt_sameprefix?versionId=Mvsc4FgJWc6gExwSw1d6wsLrnk6wdDVa 8
Visible files: 8 / Size: 64 (64 Bytes)
1 66166.....................................:Wed 17 Mar 2021 10:34:34 AM EDT:.
(git)lena:~/proj/dandi/dandi-cli-master[gh-470]git
$> tools/s3-gc-stats --stat invisible --list s3://datalad-test0-versioned
s3://datalad-test0-versioned/1version-removed-recreated.txt?versionId=B7cP2PKGRNeRZ7EktWN82UnUaGjb9dh_ 8
s3://datalad-test0-versioned/1version-removed.txt?versionId=eZ5Hgwo8azfBv3QT7aW9dmm2sbLUY.QP 8
s3://datalad-test0-versioned/2versions-nonversioned1.txt?versionId=null 8
s3://datalad-test0-versioned/2versions-removed-recreated.txt?versionId=zwW0b567gYO3puJeLMZsOmETqowJnv6l 8
s3://datalad-test0-versioned/3versions-allversioned.txt?versionId=b.qCuh7Sg58VIYj8TVHzbRS97EvejzEl 8
s3://datalad-test0-versioned/3versions-allversioned.txt?versionId=pNsV5jJrnGATkmNrP8.i_xNH6CY4Mo5s 8
Invisible files: 6 / Size: 48 (48 Bytes)

```
</details>

**Review by yarikoptic — APPROVED** [review]
Great work @jwodder !  Works like charm. Left some notes from review for posterity

Current stats:
```
$> grep -h Size /tmp/dandiarchive-3-*                                                                                                                                                             
All files: 195975 / Size: 26793923588824 (26.8 TB)
Invisible files: 13705 / Size: 8104409130145 (8.1 TB)
Old files: 2 / Size: 68846538 (68.8 MB)
Visible files: 182270 / Size: 18689514458679 (18.7 TB)
```
from the run of
```
for k in all visible invisible old ; do tools/s3-gc-stats --stat $k --exclude s3://dandiarchive/dandiarchive/dandiarchive/data/.* --list s3://dandiarchive >| /tmp/dandiarchive-3-$k.stat; done
```

**Pull request merged**